### PR TITLE
es6 compile

### PR DESCRIPTION
**Pull request by jrieken**
This is for #65372 and makes us emit es6 without dragging in es6-libraries (where possible).

### CODE REVIEW

File: src/tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,14 @@
 		"removeComments": false,
 		"preserveConstEnums": true,
 		"sourceMap": false,
-		"outDir": "../out"
+		"outDir": "../out",
+		"target": "es6",
+		"lib": [
+			"dom",
+			"es5",
+			"es2015.iterable",
```
Comment (jrieken) [inline src/tsconfig.json]: @DanielRosenwasser @mjbvz I cannot get this to work with `es2015.iterable`. Without it all destructing and for-of is illegal, independent of the `downlevelIteration`-switch. The downside is that `es2015.iterable` drags in all of `Symbol` which we don't want (in the monaco editor and below)

File: src/tsconfig.json [hunk 2]
```diff
@@ -4,7 +4,14 @@
...
+			"webworker.importscripts"
+		]
```
Comment (jrieken) [inline src/tsconfig.json]: This isn't in the schema but exactly what we want 🤓

**Comment by jrieken** [comment]
@rbuckton you mentioned that you have a `__extends`-util that can make a function extend a class. Could you share that with us? Today, all extends that extend our api class-types are broken.

**Comment by jrieken** [comment]
@rbuckton you mentioned that you have a `__extends`-util that can make a function extend a class. Could you share that with us? Today, all extends that extend our api class-types are broken.

**Comment by jrieken** [comment]
A slightly scary thing that I have encounter is this

```ts
let a:[] = undefined;
let b = [1,2,3]

let c = [...b, ...a]
```

This snippet above fails to run (correctly) when emitting es6-code but runs when emitting es5-code (which uses concat). That will surely uncover a few unpleasant surprises...

**Comment by mjbvz** [comment]
@jrieken I think strict null checks may help catch some of these case. With strict mode enabled, this code produces an error:

```js
let a: any[] | undefined = undefined;
let b = [1, 2, 3]

let c = [...b, ...a]
```

**Comment by rbuckton** [comment]
@jrieken: Unfortunately, its not `__extends` that is the problem, but rather how we transform the call to `super()` in an ES5 class that extends and ES6 class. The issue is that our emit for ES5-downlevel classes uses `_super.call`, however ES6 classes are spec'd to throw when their internal `[[Call]]` slot is invoked.

We've discussed changing our down-level emit to support this hybrid approach but that wouldn't help existing packages unless they updated to a version of TypeScript that included that change and recompiled.

Another option would be to use `--experimentalDecorators` in your API projects and use a decorator to wrap your class in something callable:

```js
function es5ClassCompat(target) {
    function _() { return Reflect.construct(target, arguments, this.constructor); }
    Object.defineProperty(_, "name", Object.getOwnPropertyDescriptor(target, "name"));
    Object.setPrototypeOf(_, target);
    Object.setPrototypeOf(_.prototype, target.prototype);
    return _;
}
...
@es5ClassCompat
class ES6NativeClass {
}
```

File: src/tsconfig.json (hunk 1 above)
Comment (DanielRosenwasser) [reply src/tsconfig.json]: Quick conservative workaround:

```ts
declare var Symbol: {
  iterator: symbol;
};
```

Comment (rbuckton) [reply src/tsconfig.json]: Is the Monaco editor also being compiled to ES6 or does it have its own tsconfig.json file?

Comment (rbuckton) [reply src/tsconfig.json]: When you use `--target es6` we use native destructuring and iteration, which requires `Symbol.iterator` at runtime, therefore we type-check that the array being destructured/iterated has a `[Symbol.iterator]()` method that can be used.

If you wan to leave this out of the tsconfig.json file, one option is to add a `/// <reference lib="es2015.iterable" />` to the files that need iteration.

Comment (jrieken) [reply src/tsconfig.json]: @rbuckton we do have a separate `tsconfig.monaco.json` which targets ES5.

**Comment by jrieken** [comment]
> re https://github.com/Microsoft/vscode/pull/68121#issuecomment-461557086

Thanks @rbuckton!

**Comment by Astrantia** [comment]
@jrieken does this improve startup performance?

**Pull request merged**